zjit/src/hir.rs

```diff
@@ -4116,43 +4158,99 @@ pub fn iseq_to_hir(iseq: *const rb_iseq_t) -> Result<Function, ParseError> {
+                target: BranchEdge { target: target_block, args: entry_state.as_args(self_param) },
+            });
+        }
+    }
+}
+
+/// Compile a jit_entry_block
+fn compile_jit_entry_block(fun: &mut Function, jit_entry_idx: usize, target_block: BlockId) {
+    let jit_entry_block = fun.jit_entry_blocks[jit_entry_idx];
+    fun.push_insn(jit_entry_block, Insn::EntryPoint { jit_entry_idx: Some(jit_entry_idx) });
+
+    // Prepare entry_state with basic block params
+    let (self_param, entry_state) = compile_entry_state(fun, jit_entry_block);
+
+    // Jump to target_block
+    fun.push_insn(jit_entry_block, Insn::Jump(BranchEdge { target: target_block, args: entry_state.as_args(self_param) }));
+}
+
+/// Compile params and initial locals for an entry block
+fn compile_entry_state(fun: &mut Function, entry_block: BlockId) -> (InsnId, FrameState) {
     let iseq = fun.iseq;
-    let mut entry_state = FrameState::new(iseq);
+    let param_size = unsafe { get_iseq_body_param_size(iseq) }.as_usize();
+
     let self_param = fun.push_insn(entry_block, Insn::Param { idx: SELF_PARAM_IDX });
-    for local_idx in 0..unsafe { get_iseq_body_param_size(iseq) }.as_usize() {
-        entry_state.locals.push(fun.push_insn(entry_block, Insn::Param { idx: local_idx + 1 })); // +1 for self
+    let mut entry_state = FrameState::new(iseq);
+    for local_idx in 0..num_locals(iseq) {
+        if local_idx < param_size {
+            entry_state.locals.push(fun.push_insn(entry_block, Insn::Param { idx: local_idx + 1 })); // +1 for self
+        } else {
+            entry_state.locals.push(fun.push_insn(entry_block, Insn::Const { val: Const::Value(Qnil) }));
+        }
     }
+    (self_param, entry_state)
+}
 
-    // Jump to the first_block. Conditionally side-exit for interpreter entry with optional arguments.
-    // TODO: Compile gen_entry_params() for !jit_entry here
-    let opt_num = unsafe { get_iseq_body_param_opt_num(iseq) };
-    if !jit_entry && opt_num > 0 {
-        let pc = fun.push_insn(entry_block, Insn::LoadPC);
-        let no_opts_pc = fun.push_insn(entry_block, Insn::Const { val: Const::CPtr(unsafe { rb_iseq_pc_at_idx(iseq, 0) } as *const u8) });
-        let test_id = fun.push_insn(entry_block, Insn::IsBitEqual { left: pc, right: no_opts_pc });
-
-        // Jump to the first_block if no optional arguments are supplied
-        fun.push_insn(entry_block, Insn::IfTrue {
-            val: test_id,
-            target: BranchEdge { target: first_block, args: entry_state.as_args(self_param) },
-        });
-
-        // Side-exit if any optional argument is supplied
-        let state = fun.push_insn(entry_block, Insn::Snapshot { state: FrameState::new(iseq) }); // no need to spill params
-        fun.push_insn(entry_block, Insn::SideExit { state, reason: SideExitReason::OptionalArgumentsSupplied });
+/// Set param_types of the function. They are copied to the param types of entry blocks.
+fn set_param_types(fun: &mut Function) {
```

Comment:
This seems to be duplicate code with Function::set_param_types. Let's keep just the method.

Reply:
Good catch https://github.com/ruby/ruby/pull/14653/commits/894b8de6138da9c098ed6b4ee7e62b1dc1d83b1a. I'm kind of surprised that Rust doesn't warn me on it while it's unused and not even pub, but I guess it's due to `![allow(dead_code)]` sitting at lib.rs.